libbeat/reader/debug/debug.go

```diff
@@ -0,0 +1,178 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package debug
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	offsetStart        = 100
+	offsetEnd          = 100
+	defaultMinBuffer   = 16 * 1024
+	defaultMaxFailures = 100
+)
+
+type state int
+
+const (
+	initial state = iota
+	running
+	stopped
+)
+
+// CheckFunc func receive a slice of bytes and returns true if it match the predicate.
+type CheckFunc func(offset int64, buf []byte) bool
+
+// Reader is a debug reader used to check the values of specific bytes from an io.Reader,
+// Is is useful is you want to detect if you have received garbage from a network volume.
+type Reader struct {
+	log           *logp.Logger
+	reader        io.Reader
+	buffer        bytes.Buffer
+	minBufferSize int
+	maxFailures   int
+	failures      int
+	predicate     CheckFunc
+	state         state
+	offset        int64
+}
+
+// NewReader returns a debug reader.
+func NewReader(
+	log *logp.Logger,
+	reader io.Reader,
+	minBufferSize int,
+	maxFailures int,
+	predicate CheckFunc,
+) (*Reader, error) {
+	return &Reader{
+		log:           log,
+		minBufferSize: minBufferSize,
+		reader:        reader,
+		maxFailures:   maxFailures,
+		predicate:     predicate,
+	}, nil
+}
+
+// Read will proxy the read call to the original reader and will periodically checks the values of
+// bytes in the buffer.
+func (r *Reader) Read(p []byte) (int, error) {
+	if r.state == stopped {
+		return r.reader.Read(p)
+	}
+
+	if r.state == running && r.failures > r.maxFailures {
+		// cleanup any remaining bytes in the buffer.
+		if r.buffer.Len() > 0 {
+			r.predicate(r.offset, r.buffer.Bytes())
+			r.buffer = bytes.Buffer{}
+		}
+		r.log.Info("Stopping debug reader, max execution reached")
+		r.state = stopped
+		return r.reader.Read(p)
+	}
+
+	if r.state == initial {
+		r.log.Infof(
+			"Starting debug reader with a buffer size of %d and max failures of %d",
+			r.minBufferSize,
+			r.maxFailures,
+		)
+		r.state = running
+	}
+
+	n, err := r.reader.Read(p)
+
+	if n != 0 {
+		r.buffer.Write(p[:n])
+		if r.buffer.Len() >= r.minBufferSize {
+			if r.failures < r.maxFailures && r.predicate(r.offset, r.buffer.Bytes()) {
+				r.failures++
+			}
+			r.buffer = bytes.Buffer{}
```

Comment:
here `r.buffer.Reset()` should be good enough. By reusing the buffer we generate less garbage and require less allocations. We can still clear if `r.buffer.Cap() > 2 * max(len(p))` (as len(p) normally is a constant as well...). Anyways, as this is just for debugging, I don't care if/when we clear the complete buffer with `Reset()` vs "zero value" here.